### PR TITLE
pulseaudio: fix PaPulseAudio_Initialize unlocking

### LINKED ISSUE #848: PaPulseAudio_Initialize does not unlock correctly on error path
This PR fixes an access-after free when calling `PaPulseAudio_UnLock`

https://github.com/PortAudio/portaudio/pull/847

However the bigger issue is that `PaPulseAudio_UnLock` is not called consistently. Either `PaPulseAudio_UnLock` needs to be called before or after `goto error` but not both or neither. On line 563 it gets locked, but there is also a `goto error` prior to that on line 550. Some later `goto error` paths unlock, some don't.

One possible fix is to have an `error_unlock:` label prior to the `error` label that is used for paths that should unlock.

### PR DESCRIPTION
**Pull request by illuusio**
When PaPulseAudio_Initialize fails it does not currently do it very well as bug report  #848 tells. Now Only free lock if it's initialized enough to be unlocked.

Unlocking is done after allocations are done. That is because Locking is just after allocations and next `goto error` should have allocated memory and drop to that. Also now Group freeing code is one place.

Resolves #848

### CODE REVIEW

File: src/hostapi/pulseaudio/pa_linux_pulseaudio.c [hunk 1]
```diff
@@ -750,10 +755,8 @@ PaError PaPulseAudio_Initialize( PaUtilHostApiRepresentation ** hostApi,
     {
         if( pulseaudioHostApi->allocations )
```
Comment (philburk) [inline src/hostapi/pulseaudio/pa_linux_pulseaudio.c:756]:
```suggestion
        if( lockTaken )
```

**Review by philburk — CHANGES_REQUESTED** [review]

**Review by RossBencina — COMMENTED** [review]
Just a suggestion...

File: src/hostapi/pulseaudio/pa_linux_pulseaudio.c (hunk 1 above)
Comment (RossBencina) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio.c:756]: I like this. It is clearer.

add
```
    int lockTaken = 0;
```

to top of function, and

```
    lockTaken = 1;
```

after the call to `PaPulseAudio_Lock`

File: src/hostapi/pulseaudio/pa_linux_pulseaudio.c [hunk 2]
```diff
@@ -211,13 +211,18 @@ void PaPulseAudio_Free( PaPulseAudio_HostApiRepresentation * ptr )
...
     {
         pa_threaded_mainloop_free( ptr->mainloop );
         ptr->mainloop = NULL;
     }
 
+    if( ptr->allocations )
+    {
+        PaUtil_FreeAllAllocations( ptr->allocations );
+        PaUtil_DestroyAllocationGroup( ptr->allocations );
+    }
```
[the remark this answers is not in the document]
Comment (illuusio) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio.c:225]: Ok i've added this one too

File: src/hostapi/pulseaudio/pa_linux_pulseaudio.c (hunk 1 above)
Comment (illuusio) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio.c]: Ok this should be also resolved

**Comment by kleinerm** [comment]
@illuusio Changes look good to me. I also gave it a quick test and it worked.

File: src/hostapi/pulseaudio/pa_linux_pulseaudio.c [hunk 3]
```diff
@@ -589,6 +597,7 @@ PaError PaPulseAudio_Initialize( PaUtilHostApiRepresentation ** hostApi,
                 break;
             case PA_CONTEXT_TERMINATED:
             case PA_CONTEXT_FAILED:
+                result = paUnanticipatedHostError;
```
Comment (RossBencina) [inline src/hostapi/pulseaudio/pa_linux_pulseaudio.c:600]: I think there's a missing call to `PaUtil_SetLastHostErrorInfo` here. perhaps separate calls for each case.

Comment (RossBencina) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio.c:600]: It's fine to do additional cleanup later, but this PR adds a `paUnanticipatedHostError`. Our policy is to not accept PRs that introduce new problems, so we'd appreciate it if you could fix this case in this PR please.

Comment (illuusio) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio.c:607]: Ok now it should be solved

Comment (RossBencina) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio.c:607]: Thanks, but now the first parameter to `PA_PULSEAUDIO_SET_LAST_HOST_ERROR` is not correct. It should be the native error code definitely not a PortAudio error code. For this PR, you could just set it to 0 if you don't have anything better. For further info check the definition of `PA_PULSEAUDIO_SET_LAST_HOST_ERROR` in `pa_linux_pulseaudio_internal.h` and then the definition of `PaUtil_SetLastHostErrorInfo` in `pa_util.h`.

Also, `initiliaze` is mis-spelled on line 606.

While I'm here, note that you'll need to fix the `PA_PULSEAUDIO_SET_LAST_HOST_ERROR` in a later PR to pass the correct host api type.

Comment (illuusio) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio.c:607]: I've correct the errors and spelling and actually hostapi type is fixed in another PR but I can fix it here if it's needed.

Comment (illuusio) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio.c]: I've removed those new error codes and now it behaves as it was before only it goes to error (problem is that it just return `PaNoError` but as it did worked like that lately it should be ok.

**Review by RossBencina — APPROVED** [review]

**Review by philburk — APPROVED** [review]

**Pull request merged**